flake.nix

```diff
@@ -1,139 +1,55 @@
-            "-DCMAKE_C_FLAGS=-D__ARM_FEATURE_DOTPROD=1"
-            "-DLLAMA_METAL=ON"
-          ] else [
-            "-DLLAMA_BLAS=ON"
-            "-DLLAMA_BLAS_VENDOR=OpenBLAS"
-          ]);
-        };
-        packages.opencl = pkgs.stdenv.mkDerivation {
-          inherit name src meta postPatch nativeBuildInputs postInstall;
-          buildInputs = with pkgs; buildInputs ++ [ clblast ];
-          cmakeFlags = cmakeFlags ++ [
-            "-DLLAMA_CLBLAST=ON"
-          ];
-        };
-        packages.cuda = pkgs.stdenv.mkDerivation {
-          inherit name src meta postPatch nativeBuildInputs postInstall;
-          buildInputs = with pkgs; buildInputs ++ [ cudatoolkit_joined ];
-          cmakeFlags = cmakeFlags ++ [
-            "-DLLAMA_CUBLAS=ON"
-          ];
-        };
-        packages.rocm = pkgs.stdenv.mkDerivation {
-          inherit name src meta postPatch nativeBuildInputs postInstall;
-          buildInputs = with pkgs.rocmPackages; buildInputs ++ [ clr hipblas rocblas ];
-          cmakeFlags = cmakeFlags ++ [
-            "-DLLAMA_HIPBLAS=1"
-            "-DCMAKE_C_COMPILER=hipcc"
-            "-DCMAKE_CXX_COMPILER=hipcc"
-            # Build all targets supported by rocBLAS. When updating search for TARGET_LIST_ROCM
-            # in github.com/ROCmSoftwarePlatform/rocBLAS/blob/develop/CMakeLists.txt
-            # and select the line that matches the current nixpkgs version of rocBLAS.
-            "-DAMDGPU_TARGETS=gfx803;gfx900;gfx906:xnack-;gfx908:xnack-;gfx90a:xnack+;gfx90a:xnack-;gfx940;gfx941;gfx942;gfx1010;gfx1012;gfx1030;gfx1100;gfx1101;gfx1102"
-          ];
-        };
-        apps.llama-server = {
-          type = "app";
-          program = "${self.packages.${system}.default}/bin/llama-server";
-        };
-        apps.llama-embedding = {
-          type = "app";
-          program = "${self.packages.${system}.default}/bin/embedding";
-        };
-        apps.llama = {
-          type = "app";
-          program = "${self.packages.${system}.default}/bin/llama";
-        };
-        apps.quantize = {
-          type = "app";
-          program = "${self.packages.${system}.default}/bin/quantize";
-        };
-        apps.train-text-from-scratch = {
-          type = "app";
-          program = "${self.packages.${system}.default}/bin/train-text-from-scratch";
-        };
-        apps.default = self.apps.${system}.llama;
-        devShells.default = pkgs.mkShell {
-          buildInputs = [ llama-python ];
-          packages = nativeBuildInputs ++ osSpecific;
-        };
-        devShells.extra = pkgs.mkShell {
-          buildInputs = [ llama-python-extra ];
-          packages = nativeBuildInputs ++ osSpecific;
-        };
-      });
+
+  outputs = { self, nixpkgs }:
+
+  let
+    inherit (nixpkgs.lib) genAttrs;
+
+    overlay = import ./.devops/overlay.nix;
+
+    importNixpkgs = system: import nixpkgs {
+      inherit system;
+      overlays = [ overlay ];
+    };
```

Comment:
H'm. Note that in the current form `github:ggerganov/llama.cpp#cuda` remains "broken" because the licenses have to be accepted at the nixpkgs import/instantiation time. I.e. you'd have to reimport nixpkgs with `config.allowUnfree = true`* for `packages.${system}.cuda` to make it work "out of the box" and in the pure evaluation mode

`*`: if you wanted to keep it precise, you could instead use something like `config.allowUnfreePredicate = p: builtins.all (license: license.free || builtins.elem license.shortName [ "CUDA EULA" "cuDNN EULA" ]) (p.meta.licenses or [p.meta.license])`.

I'd also import nixpkgs with `config.cudaSupport = true` to ensure that the dependencies (e.g. openmpi depends on `ucx` and `ucc`) use CUDA consistently as well.

I'd rather not postpone these two until later PRs.

In future we could also consider issues like those described in https://zimbatm.com/notes/1000-instances-of-nixpkgs, but right now that's strictly out of scope

Reply:
I rewrote this section of the flake (and removed the overlays that were pre-customized.)  It both appears to build and to test.